### PR TITLE
Updating MNIST Linear to TF2

### PR DESCRIPTION
**Pull request by danideter**
Ok, I think I finally got the username right this time :)
I ran into the same trouble as you did, Evan.

Anyway, can I please trouble you for your expert reviewing skills?
Many thanks,
Danielle

### CODE REVIEW

File: courses/machine_learning/deepdive/08_image/mnist_linear.ipynb [hunk 1]
```diff
@@ -9,9 +9,7 @@
    "source": [
```
Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>I'd add the following line here to make sure we are all using the same TF 2.0 version:</p><p> </p><p>!pip freeze | grep tf-nightly-2.0-preview || pip install tf-nightly-2.0-preview</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>I'd split the two lines</p><p> </p><p>print(x_train[0].shape)</p><p>tf.print(tf.size(tf.unique(y_train).y))</p><p> </p><p>into their own cell and prefix it with a markdown cell explaining what we are doing (especially for the last line)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Actually: You can just print these variables and delete the cell above (see previous comment) since the variable names are explanation enough</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <ul><li>Use print(smth, smth, smth) instead of print(smth + " " + smth + " " + smth)</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <ul><li>Remove the period in the title</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>From <a href="http://pep8online.com/" target="_blank">http://pep8online.com</a></p><p><strong>E261 </strong>at least two spaces before inline comment</p><p><strong>E262 </strong>inline comment should start with '# '</p><p> </p><p>This cell feels confusing, since it's not clear how to use this function for training.</p><p>If we keep it, we should have also the code to be able to train it (training loop from scratch with</p><p>tf.GrandientTape to compute gradients?). Personally, I'd remove it altogether, otherwise the </p><p>training code can be adapted from: <a href="https://github.com/GoogleCloudPlatform/training-data-analyst/blob/ml_on_gcp/courses/machine_learning/intro_to_tensorflow/1_core_tensorflow.ipynb" target="_blank">https://github.com/GoogleCloudPlatform/training-data-analyst/blob/ml_on_gcp/courses/machine_learning/intro_to_tensorflow/1_core_tensorflow.ipynb</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <ul><li>For readability &amp; simplicity, I'd import the layers and model (Flatten, Dense, Sequential etc.) at the beginning of the notebook, allowing us to have super code to describe the model. I'd also add to set the "input_shape" parameter to the Flatten layer so that we can then do a "model.summary()" from what this function returns. This would look something like:</li></ul><p> </p><p>&nbsp;&nbsp;model = Sequential([</p><p>&nbsp;&nbsp;&nbsp;&nbsp;Flatten(input_shape=[HEIGHT, WIDTH]),</p><p>&nbsp;&nbsp;&nbsp;Dense(NCLASSES),</p><p>&nbsp;&nbsp;&nbsp;&nbsp;Softmax()</p><p>&nbsp;&nbsp;])</p><p> </p><ul><li>I am not understanding the comment</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>In create_input_fn do we really need to return a function that returns a dataset? Doesn't Keras model.fit also work if we pass the dataset instance directly (as opposed to a function that returns the instance)? If so, we should do that instead, since it's simpler - and rename the function accordingly then.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>This is interesting. What's the advantage to use that instead of example code using the function with assert (which would be simpler to read and convey more easily the information on how to use the code to the reader as well as a mean to test it)? I am a bit afraid with that that the testing section grows as big as the notebook itself if we use the full power of the unittest suite. What do you think?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>We can also add a TensorBoard() callback to inspect the model in TB (<a href="https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/TensorBoard" target="_blank">https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/TensorBoard</a>)</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>I'd  prefer here to put directly the same test as simple assert directly on the history instance in the previous cell. It's simpler and more readable to me. What do you think?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>I though all the estimator code should be gone.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Same as previous comment</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (BenoitDherin) [inline courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Same as above. I don't think we will cover estimator anymore, won't we?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: Yeah, I got confused reading the [conversion docs](https://www.tensorflow.org/beta/guide/migration_guide#estimators). I'll go ahead and drop the estimator.

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Hmm, when I run it, the notebook is still using the beta version that comes with the Deep Learning VM. I'm a little nervous about changing the version that comes with the notebook as I've seen that result in some weird bugs. I added a line asking the user to check that they're running TF2 in case they aren't using a Deep Learning VM, and to run the above line if not. Thoughts?</p>

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Yeah, good call. That's much cleaner.</p>

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Implemented :)</p>

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Ack.</p>

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Ack.</p>

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Yeah, it's a carry over from the older lab, so I'm not surprised it seems out of place. I'm not confident it adds much value either. I'll go ahead and remove it.</p>

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Ah, good idea. The next notebook in the series expands on this, so to keep the experience consistent, I shortened it to layers.&lt;layer-name&gt;. There's lots of different layers in the next one, so there would be a lot of imports. How's that for a compromise?</p><p> </p><p>As for the comment, good call I don't understand it either. Axed!</p>

Comment (BenoitDherin) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>At some point, I believe we needed to have the tf-nightly installed in order to deploy the model. If the default TF 2.0 works with deployment, we should use that, and replace the version by the default version in the line above. </p>

Comment (BenoitDherin) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Personally, I prefer to have more complex imports and shorter/simpler code because you spend time understanding the code, not the imports ( and also the layer code can become quite long with all the parameters). Also, having imports like the following is not too ugly and gives at the very beginning of the notebook all the layers we are gonna use in the notebook as a nice compact menu:</p><p> </p><p>from tensorflow.keras.layers import (</p><p>  Dense,</p><p>  Flattern,</p><p> Softmax,</p><p>)</p><p> </p><p>your call.</p><p> </p><p>  </p>

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: Good points. I just figured out how to use docker today to deploy TF2 to ai-platform. I'll try the notebook version of TF2 (2.0.0-beta1) with it and see if it works. :)

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: Oh, that's cool. I didn't realize you could use imports that way. Learned something new :)
I'll give it a shot and see how it looks.

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: Yeah, after thinking about it, if we can pull the code from the notebook to run in automated tests, asserts would be easier to work with, huh? One advantage I like from the unit test suite is it gives a nice print out explicitly stating what passed and didn't pass. I switched to asserts and add a "Success!" print out if it passes, so hopefully best of both worlds.

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: Made it that was so it was reusable with estimator. Since that's gone now, this got simplified.

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: Done.

**Comment by danideter** [comment]
Thanks for the review, Benoit! Was I able to implement your feedback correctly?
Also remembered that we're supposed to make new folders for these, right? I went ahead a copied the new version of the lab there and reverted the old one.

**Comment by BenoitDherin** [comment]
<p>The branch we should merge into for the update is ml_on_gcp, so I changed accordingly the target branch for this PR. </p>

File: courses/machine_learning/deepdive/08_image/mnist_linear.ipynb (hunk 1 above)
Comment (BenoitDherin) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>I'd give a more concrete name to the input_fn now that this notions has kind of disappeared, like load_dataset or create_dataset. </p>

Comment (BenoitDherin) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Great! </p><p> </p><p>nit: I'd use a more concrete name (see comment above) like create_test_dataset (resp. create_train_dataset)</p>

File: courses/machine_learning/images/mnist_linear.ipynb [hunk 2]
```diff
@@ -0,0 +1,578 @@
+{
```
Comment (BenoitDherin) [inline courses/machine_learning/images/mnist_linear.ipynb]: <p>I'd add --port 8080 to make clear on which port to connect. The first time I ran this cell I got confused since it blocks the notebook execution. This is fine though because you can run the cell, look at tensorboard on localhost:8080, then interupt the cell to continue for the next cells. </p><p>Maybe add some explanation around that workflow.</p><p> </p><p>Also: I am testing the notebook locally. Does it work on dlvm? (localhost should be changed to something else up there, the port should be open also to receive connections and not blocked by firewall rules, right?)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

**Comment by BenoitDherin** [comment]
<p>This is great! I added a few very minor comments. Also I changed the branch to the one we are using for the update (ml_on_gcp). </p>

File: courses/machine_learning/deepdive/08_image/mnist_linear.ipynb (hunk 1 above)
Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: <p>Whoops, meant to do that. Good catch!</p>

Comment (danideter) [reply courses/machine_learning/deepdive/08_image/mnist_linear.ipynb]: Whoops, sorry. I meant to do that in the last commit.

File: courses/machine_learning/images/mnist_linear.ipynb (hunk 2 above)
Comment (danideter) [reply courses/machine_learning/images/mnist_linear.ipynb]: Ah, good point. I updated the instructions. Are they clearer?
Yes, I'm building and testing this code on a dlvm, which expects to use port 6006. Does that work for you locally?

**Comment by danideter** [comment]
Thanks for the review and your github magic, Benoit :)
How does it look now? Noticed that other commits got pulled into this. Is that ok?
I also added a small challenge exercise since I remembered students were asking for more of those. Does it look appropriate?
Thanks again :)

**Comment by BenoitDherin** [comment]
<p>Looks great! </p><p> </p><p>As for the bad commits that are now in danie-dev branch,  I suggest that </p><ol><li>you create a fresh  new branch of ml_on_gcp</li><li>create a new "images" folder on that new branch and copy your mnist notebook there</li><li>add, commit, push</li><li>start a PR from that new branch into ml_on_gcp</li><li>merge your work from that PR</li></ol><p> </p><p>That way we won't have to carry over these bad commits into our ml_on_gcp branch.</p>

**Comment by danideter** [comment]
Thanks Benoit! Followed your steps in #563 . Closing this out and deleting branch.
